add support for MammothModa2 model

HonestDeng opened the pull request:
<!-- markdownlint-disable -->
PLEASE FILL IN THE PR DESCRIPTION HERE ENSURING ALL CHECKLIST ITEMS (AT THE BOTTOM) HAVE BEEN CONSIDERED.

## Purpose
Resolve https://github.com/vllm-project/vllm-omni/issues/314 , add support for MammothModa2 model https://github.com/bytedance/mammothmoda
## Test Plan
Machine:
- H200(140GB) x 1

Parallel: 
- TP: None

Image:
- Size: 1024 x 1024
- DiT Step: 50

2. Image Summery

Machine:
- H200(140GB) x 1

Parallel: 
- TP: None

Image:
- Size: 1024 x 1024

## Test Result

Image in the left side is generated by  **MammothModa2 official implementation** while the right side from **vllm-omni**:
<img width="1648" height="834" alt="image" src="https://github.com/user-attachments/assets/69b6dee1-5289-4d67-a847-16f3d0964efb" />

This table shows performance in two implementations:
Stages | official-impl | vllm-omni
-- | -- | --
AR stage | 83.529s | 74.06s
DiT stage | 10.320s | 9.65s

Transfer time: 4.012ms

We get better performance.




---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING, PLEASE READ <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>** (anything written below this line will be removed by GitHub Actions)

HonestDeng commented:
@hsliuustc0106
Now we can generate an 1024x1024 pixels image, with spending 77s in AR stage and 12s in DiT.

AR Stage generates 64 * 65 = 4160 image tokens in 77s, which is slow. I found maybe the reason is that cudagraph in AR stage is not enabled in default. 

Did vllm-omni support cudagraph? Should I use cudagraph in this PR?

HonestDeng commented:
> @ZJY0516 @SamitHuang where do you think this model should be placed, under diffusion or the omni workflow?

I think this model should be placed in omni workflow, becase MammothModa2 can generate text token, like qwen25omni and qwen3omni.

hsliuustc0106 commented:
> > @ZJY0516 @SamitHuang where do you think this model should be placed, under diffusion or the omni workflow?
> 
> I think this model should be placed in omni workflow, becase MammothModa2 can generate text token, like qwen25omni and qwen3omni.

it feels like it's also similar to Bagel&HunyuanImage #203 #209

ZJY0516 commented:
> @ZJY0516 @SamitHuang where do you think this model should be placed, under diffusion or the omni workflow?

AR is computational bottleneck for this model. We should not put it in diffusion part entirely.

We can try to put diffusion stage under diffusion engine if necessary.

hsliuustc0106 commented on vllm_omni/diffusion/diffusion_engine.py:
```diff
@@ -36,13 +36,15 @@ class BackgroundResources:
 
```
this is not related to your PR

HonestDeng replied:
I have just figured out why I modified this un-related file. It's an `merge accident`.

Three days ago, I tried to merge main branch to my working branch and there were some conflicts.

Maybe because my wrong operations, many modifiction by other people from main branch got into my branch.

I'll revert this accident modification.

hsliuustc0106 commented on tests/entrypoints/openai/test_image_server.py:
```diff
@@ -0,0 +1,439 @@
+# SPDX-License-Identifier: Apache-2.0
```
why you add this? the image_server is already done in another PR

hsliuustc0106 commented on examples/offline_inference/mammothmodal2_preview/mammoth_moda2_image_summary.yaml:
```diff
@@ -0,0 +1,18 @@
+stage_args:
```
here do we only have one stage?

HonestDeng commented on vllm_omni/entrypoints/omni_stage.py:
```diff
@@ -436,7 +436,9 @@ def _stage_worker(
 
     # Sequential initialization on the same device to avoid memory calculation errors
     # when multiple instances start simultaneously
-    # For TP/PP/DP/SP, we need to lock ALL devices that will be used by this stage
```
Sorry, I'll revert this.

HonestDeng commented:
> do we have any test result for performance and accuracy?

I'm working on this.

HonestDeng commented:
Image in the left side is generated by  **MammothModa2 official implementation** while the right side from **vllm-omni**:
<img width="1670" height="836" alt="image" src="https://github.com/user-attachments/assets/62c5c086-4bab-4d55-acea-d1186caee72b" />

运行阶段 | official-impl (耗时) | vllm-omni (耗时)
-- | -- | --
AR stage | 91.493s | 63.325s
DiT stage | 8.826s | **12.975s**

Here is detailed performance data of vllm-omni:
```
 {'type': 'request_level_metrics',
  'request_id': '0_31e7f6a6-c2be-43bc-b46a-c97b5e528ac6',
  'e2e_time_ms': 86702.4142742157,
  'e2e_tpt': 0.0,
  'num_tokens_out': 0,
  'transfers_total_time_ms': 10.648012161254883,
  'transfers_total_bytes': 1664065,
  'stages': {0: {'stage_gen_time_ms': 63325.48713684082, 'num_tokens_out': 4162},
             1: {'stage_gen_time_ms': 12975.892543792725, 'num_tokens_out': 0}}}
 [Summary] {'e2e_requests': 1,
  'e2e_total_time_ms': 86702.88419723511,
  'e2e_sum_time_ms': 86702.4142742157,
  'e2e_total_tokens': 0,
  'e2e_avg_time_per_request_ms': 86702.4142742157,
  'e2e_avg_tokens_per_s': 0.0,
  'wall_time_ms': 86702.88419723511,
  'final_stage_id': {'0_31e7f6a6-c2be-43bc-b46a-c97b5e528ac6': 1},
  'stages': [{'stage_id': 0,
              'requests': 1,
              'tokens': 4162,
              'total_time_ms': 73651.93319320679,
              'avg_time_per_request_ms': 73651.93319320679,
              'avg_tokens_per_s': 56.50903947194529},
             {'stage_id': 1,
              'requests': 1,
              'tokens': 0,
              'total_time_ms': 13016.989707946777,
              'avg_time_per_request_ms': 13016.989707946777,
              'avg_tokens_per_s': 0.0}],
  'transfers': [{'from_stage': 0,
                 'to_stage': 1,
                 'samples': 1,
                 'total_bytes': 1664065,
                 'total_time_ms': 4.790782928466797,
                 'tx_mbps': 2778.7775398666267,
                 'rx_samples': 1,
                 'rx_total_bytes': 1664065,
                 'rx_total_time_ms': 2.084016799926758,
                 'rx_mbps': 6387.913955620638,
                 'total_samples': 1,
                 'total_transfer_time_ms': 10.648012161254883,
                 'total_mbps': 1250.2352362481806}]}
```

hsliuustc0106 commented:
> This PR now encompasses about 5k lines of changes, which is quite large.

yes, I also doubt whether we need to copy so many codes from other repos

ZJY0516 commented on vllm_omni/model_executor/models/mammoth_moda2/mammothmoda2_dit_layer/rope_real.py:
```diff
@@ -0,0 +1,251 @@
+from torch import nn
+
+
+def apply_real_rotary_emb(x: torch.Tensor, freqs_cos: torch.Tensor, freqs_sin: torch.Tensor) -> torch.Tensor:
```
Could you please try to use rope layer in vll-omni diffusion?

HonestDeng replied:
mrope in vllm-omni is not compatible with MammothModa

ZJY0516 commented on vllm_omni/model_executor/models/mammoth_moda2/mammothmoda2_dit_layer/rope_real.py:
```diff
@@ -0,0 +1,251 @@
+    ntk_factor: float = 1.0,
+    freqs_dtype: torch.dtype = torch.float32,
+):
+    # 复用 diffusers 的实现，避免在仓库内维护重复代码。
```
remove chinese comments

hsliuustc0106 commented:
> > * [ ]   add docs supported models
> > * [ ]   add examples
> 
> You mean I need to finish this two tasks before merging?

yes

hsliuustc0106 commented on vllm_omni/worker/gpu_model_runner.py:
```diff
@@ -649,13 +649,15 @@ def _gather_runtime_additional_information(self) -> list[dict]:
         per_req_runtime_info = []
         for req_id in self.input_batch.req_ids:
             req_state = self.requests.get(req_id)
+            generated_len = len(req_state.output_token_ids) if req_state is not None else 0
```
why we need to add this? 
@tzhouam PTAL

HonestDeng replied:
When generating an image, Mammothmoda split an image into a grid, say 32x32 token. And in the end of one row, Mammothmoda inserts an "eol"(end-of-line) token explicitly. Therefore, `generated_len` is required for the generation process to determine when to insert an "eol" token.

Here is how Mammothmoda implements:
https://github.com/bytedance/mammothmoda/blob/768cd4e48ad76ae6b5e9305c2a6adbec2578f2a7/mammothmoda2/model/modeling_mammothmoda2.py#L308

hsliuustc0106 commented:
update your test plan and results
is this model AR heavy or dit heavy?
what's the transfer time between AR and dit?

HonestDeng commented:
OK. I'll move dit-stage code to diffusion folder after https://github.com/vllm-project/vllm-omni/pull/726 merged.

lishunyang12 commented:
@HonestDeng Hey, the MammothModa2 implementation looks pretty complete — AR + DiT stages, tested on H200 for both T2I and image summary. I see you're also touching gpu_generation_model_runner.py and gpu_model_runner.py — are those changes compatible with the current codebase, or does this need a rebase? Would be great to get this merged.

HonestDeng commented:
Thanks for reviewing. I've fixed the these 5 issues. And it is strange for the first issue: answer_start_index = max(L - 10, 0).

Actually, I just followed the [original project](https://github.com/bytedance/mammothmoda/blob/240130f3c26652c07f67eba48bea6afd78f70c35/mammothmoda2/utils/t2i_utils.py#L310) to use the last 10 token. I didn't figure out the reason. It's quite weird. 

Anyway, I've fixed the problem by using all generated token as answer.

HonestDeng commented:
I've run the test_mammoth_moda2.py in my local machine and pass all the test cases.

princepride commented:
> I've run the test_mammoth_moda2.py in my local machine and pass all the test cases. Is it OK?

Thanks! I will review it tmr.

princepride commented on examples/offline_inference/mammothmodal2_preview/image.png:
Plz remove it

princepride commented on vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2_dit.py:
I think we can directly use dit model under diffusion folder.

princepride commented on vllm_omni/model_executor/models/mammoth_moda2/mammoth_moda2_ar.py:
Typically, in vLLM, we put processor and model implement in same file, please combine mammoth_moda2_ar.py and mammoth_moda2.py

hsliuustc0106 commented:
> Thanks for reviewing. I've fixed the these 5 issues. And it is strange for the first issue: answer_start_index = max(L - 10, 0).
> 
> Actually, I just followed the [original project](https://github.com/bytedance/mammothmoda/blob/240130f3c26652c07f67eba48bea6afd78f70c35/mammothmoda2/utils/t2i_utils.py#L310) to use the last 10 token. I didn't figure out the reason. It's quite weird.
> 
> Anyway, I've fixed the problem by using all generated token as answer.

you can comment an issue at the original project

hsliuustc0106 commented:
> I've run the test_mammoth_moda2.py in my local machine and pass all the test cases. Is it OK?

any performance speed up update？

princepride commented on vllm_omni/model_executor/stage_configs/mammoth_moda2_ar.yaml:
Thanks for your clarify.

princepride commented:
```
python3 examples/offline_inference/mammothmodal2_preview/run_mammothmoda2_t2i.py   --model bytedance-research/MammothModa2-Preview --stage-c
onfig ./vllm_omni/model_executor/stage_configs/mammoth_moda2.yaml   --prompt "A stylish woman riding a motorcycle 
in NYC, movie poster style"   --height 1024   --width 1024   --num-inference-steps 50   --text-guidance-scale 4.0 
  --out output.png
/proj-tango-pvc/users/zhipeng.wang/workspace/vllm-omni/vllm_omni/__init__.py:32: RuntimeWarning: Failed to import version from _version.py: No module named 'vllm_omni._version'
This typically happens in development mode before building.
Using fallback version 'dev'.
  from .version import __version__, __version_tuple__  # isort:skip
Traceback (most recent call last):
  File "/proj-tango-pvc/users/zhipeng.wang/workspace/vllm-omni/examples/offline_inference/mammothmodal2_preview/run_mammothmoda2_t2i.py", line 241, in <module>
    main()
  File "/proj-tango-pvc/users/zhipeng.wang/workspace/vllm-omni/examples/offline_inference/mammothmodal2_preview/run_mammothmoda2_t2i.py", line 186, in main
    gen_cfg = load_t2i_generation_config(args.model)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/proj-tango-pvc/users/zhipeng.wang/workspace/vllm-omni/examples/offline_inference/mammothmodal2_preview/run_mammothmoda2_t2i.py", line 57, in load_t2i_generation_config
    raise FileNotFoundError(f"Config not found: {gen_cfg_path}")
FileNotFoundError: Config not found: bytedance-research/MammothModa2-Preview/t2i_generation_config.json
```

I got an error when using huggingface id, PTAL.

princepride reviewed (changes requested):
PTAL

HonestDeng commented:
Now huggingface id has been supported and pre-commit run successfully.

HonestDeng commented:
> Roughly LGTM, after testing, I will approve it.

Thanks
